### PR TITLE
[SYCL] [ON_HOLD] Add sycl:: math functions to the fpbuiltins.

### PR DESCRIPTION
**Pull request by zahiraam**
Add the sycl:: math functions to the fpbuiltin so that we can generate FPaccuracy for them.

### CODE REVIEW

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (andykaylor) [inline clang/lib/CodeGen/CGBuiltin.cpp:22885]: I think this should just translate to the regular fpbuiltin intrinsic. The llvm-spirv translator will handle translating that back to the __spirv_ocl_ functions when needed.

**Review by andykaylor — APPROVED** [review]
This looks right to me. We better talk to the OpenCL and SYCL people to make sure they are OK with it.

**Comment by zahiraam** [comment]
> This looks right to me. We better talk to the OpenCL and SYCL people to make sure they are OK with it.
Thanks @andykaylor . I can open up the patch for review. Who do you suggest I should add from OpenCL/SYCL besides the default reviewers here, to make sure they agree with the changes?

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (premanandrao) [inline clang/lib/CodeGen/CGBuiltin.cpp:23053]: For my own curiosity, how did you (or how can one) convince yourself that the entries are complete?   Is there a list of functions you were working from?

File: clang/test/CodeGen/fp-accuracy.c
Comment (premanandrao) [inline clang/test/CodeGen/fp-accuracy.c:399]: Should have caught this earlier, but please add a top-level comment explaining what this test does.

File: clang/test/CodeGen/fp-accuracy.c
Comment (premanandrao) [inline clang/test/CodeGen/fp-accuracy.c:453]: The other attributes seem self-explanatory, but if you could add a comment for what the 'fpbuiltin-max-error"="1.0"' and other values mean, that would be helpful to those looking at this test later.

**Review by premanandrao — COMMENTED** [review]
Just some nits.

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (zahiraam) [reply clang/lib/CodeGen/CGBuiltin.cpp:23053]: There is no list AFAIK, and there are probably still some functions (especially the single precision ones) that are still missing.

**Review by smanna12 — APPROVED** [review]

**Comment by zahiraam** [comment]
@intel/llvm-gatekeepers Can this patch be merged in? The fail is not related to my patch. Thanks.

**Comment by dm-vodopyanov** [comment]
> @intel/llvm-gatekeepers Can this patch be merged in? The fail is not related to my patch. Thanks.

Waiting for review from @intel/llvm-reviewers-runtime:

image

File: sycl/test/basic_tests/fp-accuracy.cpp
Comment (aelovikov-intel) [inline sycl/test/basic_tests/fp-accuracy.cpp:24]:
```suggestion
  buffer<float, 1> out(1);
```

File: sycl/test/basic_tests/fp-accuracy.cpp
Comment (aelovikov-intel) [inline sycl/test/basic_tests/fp-accuracy.cpp:14]:
```suggestion
```

File: sycl/test/basic_tests/fp-accuracy.cpp
Comment (aelovikov-intel) [inline sycl/test/basic_tests/fp-accuracy.cpp:35]:
```suggestion
    sycl::accessor output{out, cgh, sycl::write_only};
```

same in line 43.

File: sycl/test/basic_tests/fp-accuracy.cpp
Comment (aelovikov-intel) [inline sycl/test/basic_tests/fp-accuracy.cpp:53]: Are we missing a `CHECK-LABEL` for `Kernel3` here? If not, do we even need the kernel?

**Comment by aelovikov-intel** [comment]
> This looks right to me. We better talk to the OpenCL and SYCL people to make sure they are OK with it.

I believe @againull was recently working with FP precision stuff. @againull , any comments on this?

File: sycl/test/basic_tests/fp-accuracy.cpp
Comment (zahiraam) [reply sycl/test/basic_tests/fp-accuracy.cpp]: Added the CHECK-LABEL now for readability.

File: sycl/test/basic_tests/fp-accuracy.cpp
Comment (aelovikov-intel) [inline sycl/test/basic_tests/fp-accuracy.cpp:40]: Please update same as line 32 and drop line 11.

File: sycl/test/basic_tests/fp-accuracy.cpp
Comment (aelovikov-intel) [inline sycl/test/basic_tests/fp-accuracy.cpp:17]:
```suggestion
  float input = 42.0f;
```

just in case.

File: sycl/test/basic_tests/fp-accuracy.cpp
Comment (aelovikov-intel) [inline sycl/test/basic_tests/fp-accuracy.cpp:54]: Does this patch affect device compilation?

Comment (andykaylor) [reply sycl/test/basic_tests/fp-accuracy.cpp:53]: I'm not sure I understand your question. Different calls will be generated in the IR, so depending on the required accuracy it should result in a different implementation being selected. By the time we get to device compilation, these intrinsics will have been translated back to the __ocl_spv_* functions, but they should have an attribute attached with the accuracy requirement.

Comment (aelovikov-intel) [reply sycl/test/basic_tests/fp-accuracy.cpp:53]: I don't see any CHECKs for generated device code in this PR. We also have a FAST_MATH-based "optimization" implemented purely in our SYCL Device headers. I'd like to see a unit test that shows how this PR and the pre-existing optimization interact. Maybe we can/should drop the latter here as well.

Comment (zahiraam) [reply sycl/test/basic_tests/fp-accuracy.cpp:53]: The CHECK lines in fp-accuracy.cpp represent IR generated from the device compilation (first compilation).  That's what you want right?
Can you point me to a unit test that uses the fast-math based optimizations?

Comment (zahiraam) [reply sycl/test/basic_tests/fp-accuracy.cpp]: @aelovikov-intel Updated the run line as we discussed. Let me know if this works for you. Working on the -O2 version.

Comment (andykaylor) [reply sycl/test/basic_tests/fp-accuracy.cpp]: > We also have a FAST_MATH-based "optimization" implemented purely in our SYCL Device headers.
Can you give me more information about this? I'm not sure what it means.

Comment (aelovikov-intel) [reply sycl/test/basic_tests/fp-accuracy.cpp]: It's done inside https://github.com/intel/llvm/blob/sycl/sycl/source/builtins_generator.py (although I'm in the process of major refactoring in this area). In a nutshell, we're doing this (that is what the script generates):

```
template <typename T0>
std::enable_if_t<(detail::is_vec_or_swizzle_v<T0> &&
                  detail::is_valid_elem_type_v<T0, float, double, half> &&
                  detail::is_valid_size_v<T0, 16, 1, 2, 3, 4, 8>),
                 detail::simplify_if_swizzle_t<T0>>
sin(T0 a0) __NOEXC {
  if constexpr (detail::use_fast_math_v<T0>) {
    return __sycl_std::__invoke_native_sin<detail::simplify_if_swizzle_t<T0>>(
        detail::simplify_if_swizzle_t<T0>(a0));
  }
  return __sycl_std::__invoke_sin<detail::simplify_if_swizzle_t<T0>>(
      detail::simplify_if_swizzle_t<T0>(a0));
}
```

where `use_fast_math_v` is defined like this https://github.com/intel/llvm/blob/fab8e8ce7acc86aba4cc26bfc1775ade2c6ab1d8/sycl/include/sycl/builtins_utils_scalar.hpp#L35-L42

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (steffenlarsen) [inline clang/lib/CodeGen/CGBuiltin.cpp:22964]: Could the SYCL implementation be using other built-in functions to achieve this without needing these special rules? E.g. could we be calling another `sqrt` function than `__spirv_ocl_sqrt` and get this behavior directly without needing these new cases?

Comment (zahiraam) [reply clang/lib/CodeGen/CGBuiltin.cpp:22964]: If we want to avoid adding these new cases, then we need to have the FE generates different names (same names than std::* functions?) for the sycl math functions.
These sycl math functions are all defined in SPRIVBuiltins.td and tablegen generates those __spirv_ocl_*. But not sure if we can prefix them differently. @Andy Kaylor what do you think? 
Would you want that for all math functions or only specific ones?

File: sycl/test/basic_tests/fp-accuracy.cpp
Comment (andykaylor) [reply sycl/test/basic_tests/fp-accuracy.cpp]: Second example: https://godbolt.org/z/hdcqT7K5G

That one shows the symbol not being defined, even though most fast-math flags will be set. The reason this is important is that we're changing the behavior of "-fp-model fast" to honor NaNs and infinites, so __FAST_MATH__ won't be defined.

Comment (zahiraam) [reply sycl/test/basic_tests/fp-accuracy.cpp]: Compiling with -ffast-math actually doesn't create any fpbuiltin intrinsics. The reason being that the IR generated with 
-ffast-math for math functions is of the form __spirv_ocl_native_* and will therefore not go through this implementation. So this patch has no effect on the optimizations.

Comment (andykaylor) [reply sycl/test/basic_tests/fp-accuracy.cpp]: > Compiling with -ffast-math actually doesn't create any fpbuiltin intrinsics. The reason being that the IR generated with
> -ffast-math for math functions is of the form __spirv_ocl_native_* and will therefore not go through this implementation. So this patch has no effect on the optimizations (at least for now, if we decide to also add these here).

I think that's a problem. That's probably happening because of the header file handling that Andrei mentioned, but I would assume that if the user specifies both -ffast-math and -ffp-accuracy=high, they want the high accuracy implementations. If they use -ffast-math and -ffp-accuracy=low, they probably want the native implementations where they exist, but we should be able to make that happen during the function selection phase. That's not happening now, but it should when that's faster than a library call and meets the accuracy requirements.

For CPU I came across this with sqrt in my testing. The __svml_sqrt1_ep function isn't as accurate as it claims to be, but the Numerics team confirmed that we should be using the native instruction in that case anyway because it's faster.

Comment (aelovikov-intel) [reply sycl/test/basic_tests/fp-accuracy.cpp]: > though in the near term that won't happen based on the -ffast-math option alone. It will take some time to get optimization working with the fpbuiltin intrinsics.

In this case, I think we should keep our header-based optimization in place to avoid performance regressions. I do think that we need some better testing here, but I wouldn't object on not handling that in this PR.

**Pull request closed without merging**
